### PR TITLE
add_dataset_dataloader

### PR DESCRIPTION
**Pull request by AntonionyBul**

### CODE REVIEW

File: data_source/train/data_0.csv
Comment (kirilltobola) [inline data_source/train/data_0.csv:1]: Этот файл не нужно было в коммит добавлять)

File: data_source/train/data_1.csv
Comment (kirilltobola) [inline data_source/train/data_1.csv:1]: Этот файл не нужно было в коммит добавлять)

File: dataset/config.json [hunk 1]
```diff
@@ -0,0 +1,29 @@
+{
+	"num_labels": 170,
```
Comment (kirilltobola) [inline dataset/config.json:2]: Пока не нужен параметр

File: dataset/config.json [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+	"num_gpu": 4,
+	"save_period_in_epochs": 10,
+	"metrics": ["f1_micro", "f1_macro", "f1_weighted"],
```
Comment (kirilltobola) [inline dataset/config.json:5]: Не нужен

File: dataset/config.json [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+	"pretrained_model_name": "bert-base-multilingual-uncased",
+	"table_serialization_type": "column_wise",
```
Comment (kirilltobola) [inline dataset/config.json:7]: Не нужен

File: dataset/config.json [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+	"batch_size": 32,
+	"num_epochs": 30,
+	"random_seed": 2024,
+	"logs_dir": "logs/",
+	"train_log_filename": "logs/train.log",
+	"test_log_filename": "logs/test.log",
+	"start_from_checkpoint": false,
+	"checkpoint_dir": "checkpoints/",
+	"checkpoint_name": "model_best_f1_weighted.pt",
+	"inference_model_name": "model_table_wise.pt",
```
Comment (kirilltobola) [inline dataset/config.json:17]: Не нужен

File: dataset/config.json [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+	"inference_dir": "data/inference/",
```
Comment (kirilltobola) [inline dataset/config.json:18]: Не нужен

File: dataset/dataloader.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+import numpy as np
+import pandas
+import pandas as pd
+from torch.utils.data import DataLoader
+from torch.utils.data.sampler import SubsetRandomSampler
+from transformers import BertTokenizer, PreTrainedTokenizerBase
+
+from dataset import TableDataset
+
+class CtaDataLoader(DataLoader):
```
Comment (kirilltobola) [inline dataset/dataloader.py:12]: Предлагаю назвать `TableDataLoader`

File: dataset/dataloader.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+        Returns:
+            tuple: Train and valid random samplers.
+        """
+        if isinstance(split, int):
+            assert 0 < split < self.num_tables
+            len_valid = split
+        else:
+            len_valid = int(self.num_tables * split)
+
+        valid_mask = pd.Series(dataset["table_id"].unique()).sample(n=len_valid, random_state=14)
```
Comment (kirilltobola) [inline dataset/dataloader.py:76]: `random_state` лучше вынести куда-нибудь (например, в конфиг)

File: dataset/dataloader.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+        valid_df = dataset[dataset["table_id"].isin(valid_mask)]
+        valid_ids = valid_df.index.to_numpy()
+
+        train_df = dataset[~dataset["table_id"].isin(valid_mask)]
+        train_ids = train_df.index.to_numpy()
+
+        # valid_ids = dataset_ids[0:len_valid]
```
Comment (kirilltobola) [inline dataset/dataloader.py:83]: Комменты можно удалить

File: dataset/dataloader.py [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+
+        return DataLoader(sampler=self.valid_sampler, **self.init_kwargs)
+
+
+
+if __name__ == "__main__":
+    pass
+
+
+    # from config import Config
```
Comment (kirilltobola) [inline dataset/dataloader.py:101]: Код для теста обертки лучше оставить в этом блоке (раскомментить)

File: dataset/dataset.py [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+        self.transform = transform
+        self.target_transform = target_transform
+
+    def __len__(self):
+        return len(self.df)
+
+    def __getitem__(self, idx):
+        return {
+            "data": self.df.iloc[idx]["data"],
+            "labels": self.df.iloc[idx]["labels"],
```
Comment (kirilltobola) [inline dataset/dataset.py:66]: У нас вместо лейблов будут заголовки столбцов, поэтому тут заменить на `headers`

File: dataset/dataset.py [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+            labels = torch.LongTensor(table["label_id"].values)
+
+            data_list.append(
+                [table_id, num_cols, tokenized_columns_seq, labels]
+            )
+
+        return pd.DataFrame(
+            data_list,
+            columns=["table_id", "n_cols", "data", "labels"]
+        )
```
Comment (kirilltobola) [inline dataset/dataset.py:151]: В данном случае нужна другая стратегия токенизации, более простая.

В RuTaBERT мы токенизировали все столбцы таблицы в одну последовательность `[CLS] token_11 token_12 ... [SEP] [CLS] token_21 ... [SEP]`. 

Сейчас нам нужно просто каждый столбец таблицы токенизировать + вставить в начало столбца специальный токен `[CLS]` и `[SEP]` в конец столбца.

Вернуть в том же формате. Вместо  столбца `labels` в результирующем датафреме у нас будет заголовок, поэтому замени его название на `header`.

File: dataset/dataset.py [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+
+
+if __name__ == "__main__":
+    pass
```
Comment (kirilltobola) [inline dataset/dataset.py:155]: Добавить пример использования датасета, и ты забыл пустую строку добавить в конец файла.

**Review by kirilltobola — CHANGES_REQUESTED** [review]
1) Исправь мелкие замечания
2) перепиши в `dataset.py` функцию `_create_dataset`. Нужна другая стратегия токенизации для столбцов.

File: dataset/dataset.py (hunk 11 above)
Comment (kirilltobola) [reply dataset/dataset.py:151]: @AntonionyBul 
**UPD:** я тут подумал, нам вообще эта функция здесь не нужна. Токенизацию нужно будет делать после аугментаций. Создам сам функцию, которую повешаю потом на даталоадер.
